### PR TITLE
[forge] LatencyBreakdown counters and use them in success criteria

### PR DESCRIPTION
**Pull request by igor-aptos**
- refactor SystemMetrics to be cleaner, and build LatencyBreakdown metrics in the same flow
- have Swarm surface API for generic querying range from prometheus, and move validation and what needs to be fetched to SuccessCriteria
- move retries from system metrics alone, to prometheus calls themselves
- rename system_metrics.rs to prometheus_metrics.rs - and have fetching of system and latency metrics there
- move threshold logic to SuccessCriteria

- add fetching QS and consensus latency breakdown, and having a way to assert they pass.

### CODE REVIEW

File: testsuite/forge-cli/src/main.rs [hunk 1]
```diff
@@ -216,6 +218,24 @@ struct Resize {
...
+        MetricsThreshold::new(12.0, 30),
+        // Check that we don't use more than 5 GB of memory for 30% of the time.
+        MetricsThreshold::new_gb(5.0, 30),
+    )
+});
+static SYSTEM_12_CORES_10GB_THRESHOLD: Lazy<SystemMetricsThreshold> = Lazy::new(|| {
+    SystemMetricsThreshold::new(
+        // Check that we don't use more than 12 CPU cores for 30% of the time.
+        MetricsThreshold::new(12.0, 30),
+        // Check that we don't use more than 5 GB of memory for 30% of the time.
```
Comment (bchocho) [inline testsuite/forge-cli/src/main.rs:234]: 5 GB -> 10 GB

Also nit GB -> GiB

File: testsuite/forge/src/success_criteria.rs [hunk 2]
```diff
@@ -22,10 +28,123 @@ pub enum LatencyType {
...
+    pub fn new(cpu_threshold: MetricsThreshold, memory_threshold: MetricsThreshold) -> Self {
+        Self {
+            cpu_threshold,
+            memory_threshold,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct LatencyBreakdownThreshold {
```
Comment (bchocho) [inline testsuite/forge/src/success_criteria.rs:103]: remove `pub`

File: testsuite/forge/src/interface/prometheus_metrics.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    samples.insert(
+        LatencyBreakdownSlice::ConsensusProposalToOrdered,
+        MetricSamples::new(consensus_proposal_to_ordered_samples),
+    );
+    samples.insert(
+        LatencyBreakdownSlice::ConsensusOrderedToCommit,
+        MetricSamples::new(consensus_ordered_to_commit_samples),
+    );
+    samples.insert(
+        LatencyBreakdownSlice::ConsensusProposalToCommit,
```
Comment (bchocho) [inline testsuite/forge/src/interface/prometheus_metrics.rs:179]: Not used?

File: testsuite/forge-cli/src/main.rs [hunk 4]
```diff
@@ -1519,12 +1524,15 @@ fn realistic_env_max_load_test(
                 )
                 .add_system_metrics_threshold(SystemMetricsThreshold::new(
                     // Check that we don't use more than 12 CPU cores for 30% of the time.
-                    MetricsThreshold::new(14, max_cpu_threshold),
+                    MetricsThreshold::new(14.0, max_cpu_threshold),
```
Comment (bchocho) [inline testsuite/forge-cli/src/main.rs:1527]: Value doesn't match comment (12 vs 14)

**Review by bchocho — APPROVED** [review]
Looks great. Just some nits

**Review by sitalkedia — APPROVED** [review]
LGTM % minor nits

File: testsuite/forge-cli/src/main.rs [hunk 5]
```diff
@@ -216,6 +218,24 @@ struct Resize {
     enable_haproxy: bool,
 }
 
+// common metrics thresholds:
+static SYSTEM_12_CORES_5GB_THRESHOLD: Lazy<SystemMetricsThreshold> = Lazy::new(|| {
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply testsuite/forge-cli/src/main.rs:222]: there is no system_metrics module any more :) 

thresholds structs are defined in the SuccessCriteria, I don't think there is better. they are only used in this file, seems fine to have constants here.

**Pull request merged**